config/setup/startupshutdown_test.go

```diff
@@ -0,0 +1,56 @@
+package setup
+
+import (
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+
+	"github.com/mholt/caddy/config/parse"
+	"github.com/mholt/caddy/server"
+)
+
+// The Startup function's tests are symmetrical to Shutdown tests,
+// because the Startup and Shutdown functions share virtually the
+// same functionality
+func TestStartupWithBlockingCommand(t *testing.T) {
+	var startupFuncs []func() error
+
+	startupCommand := `-startup mkdir temp_dir_for_testing
+			   -startup mkdir temp_dir_for_testing &
+			   -startup highly_unlikely_to_exist_command 
+			  `
+	c := &Controller{
+		Config:    &server.Config{Startup: startupFuncs},
+		Dispenser: parse.NewDispenser("", strings.NewReader(startupCommand)),
+	}
+
+	_, err := Startup(c)
+	if err != nil {
+		t.Errorf("Expected no errors, got: %v", err)
+	}
+
+	tests := []struct {
+		shouldExecutionErr bool
+		shouldRemoveErr    bool
+	}{
+		{false, false},
+		{false, true},
+		{true, true},
+	}
+
+	for i, test := range tests {
+		err = c.Startup[i]()
+		if err != nil && !test.shouldExecutionErr {
+			t.Errorf("Test %d recieved an error of:\n%v", i, err)
+		}
+		err = os.Remove("./temp_dir_for_testing")
+		if err != nil && !test.shouldRemoveErr {
+			t.Errorf("Test %d recieved an error of:\n%v", i, err)
+		}
+
+	}
+
+	exec.Command("rm", "-r", "./temp_dir_for_testing").Start()
```

Comment:
The "startup mkdir temp_dir_for_testing &" command is nonblocking, therefore this command will not create the "temp_dir_for_testing" directory before the os.Remove call on line 47. I expect the presence of this error on line 47 to test whether the "startup mkdir temp_dir_for_testing &" command executed in a nonblocking manner.  In order to actually remove the "temp_dir_for_testing" directory after line 47 produces an error I included the remove command on line 54. 

The extra remove statement on line 54 is not needed for the execution of the "-startup mkdir temp_dir_for_testing" command, because this command is non-blocking; therefore, the os.Remove call on line 47 will not produce an error. 

I'll add comments right after the startupCommand variable to specify whether each command is blocking and whether the command should produce an error. Also I'll add a concise comment on line 54 to explain its use.